ocaml/xapi/xapi_xenops.ml

```diff
@@ -1844,30 +1839,34 @@ let update_vgpu ~__context id =
               (fun (_, vgpu_record) -> vgpu_record.API.vGPU_device = (snd id))
               vgpu_records
           in
-          Opt.iter
-            (fun (xenopsd_vgpu, state) ->
-               if state.plugged then begin
-                 let scheduled =
-                   Db.VGPU.get_scheduled_to_be_resident_on ~__context ~self:vgpu
-                 in
-                 if Db.is_valid_ref __context scheduled
-                 then begin
-                   Helpers.call_api_functions ~__context
-                     (fun rpc session_id ->
-                        XenAPI.VGPU.atomic_set_resident_on ~rpc ~session_id
-                          ~self:vgpu ~value:scheduled)
-                 end;
-                 if not vgpu_record.API.vGPU_currently_attached
-                 then Db.VGPU.set_currently_attached ~__context
-                     ~self:vgpu ~value:true
-               end else begin
-                 if vgpu_record.API.vGPU_currently_attached
-                 then Db.VGPU.set_currently_attached ~__context
-                     ~self:vgpu ~value:false;
-                 try Client.VGPU.remove dbg id
-                 with e -> debug "VGPU.remove failed: %s" (Printexc.to_string e)
-               end) info;
-          Xenops_cache.update_vgpu id (Opt.map snd info)
+          (* We only proceed if the VGPU is not a passthrough VGPU. In the
+           * passthrough case, the VM will have a PCI device, and update_pci
+           * will set VGPU.{resident_on;currently_attached}. *)
+          if Xapi_vgpu_type.requires_passthrough ~__context ~self:vgpu_record.API.vGPU_type = None then
```

Comment:
This code section is a quite deeply nested if-then-else chain with mixed-in side effects. Could this be made flatter? One way would be to start with a `match state.plugged, vgpu_record.API.vGPU_currently_attached` on the outermost level and using `when`. This would lead to a table-like structure that I would find easier to reason about. This is also a matter of taste and it's just a suggestion.